File: packages/wnode-sdk-ts/test/mesh/proof.test.ts
import { MeshProofAggregator } from '../../src/mesh/proof';
import { WnodeDeterminismError } from '../../src/errors';
import { WorkflowStepResult } from '../../src/mesh/types';

describe('MeshProofAggregator', () => {
  let aggregator: MeshProofAggregator;

  beforeEach(() => {
    aggregator = new MeshProofAggregator();
  });

  it('aggregates valid step proofs deterministically', () => {
    const results: WorkflowStepResult[] = [
      {
        workflowId: 'wf-1',
        stepId: 'step-2', // test sorting
        nodeId: 'node-2',
        stepHash: '0x2222',
        localProof: {
          version: '1.0',
          workflowId: 'wf-1',
          stepHashes: ['0x2222'],
          timestamp: 1000,
          chainId: 1,
          blockTag: { finalized: true }
        }
      },
      {
        workflowId: 'wf-1',
        stepId: 'step-1',
        nodeId: 'node-1',
        stepHash: '0x1111',
        localProof: {
          version: '1.0',
          workflowId: 'wf-1',
          stepHashes: ['0x1111'],
          timestamp: 1000,
          chainId: 1,
          blockTag: { finalized: true }
        }
      }
    ];

    const proof = aggregator.aggregateProofs('wf-1', results, 1, { finalized: true });
    expect(proof.version).toBe('1.0');
    expect(proof.workflowId).toBe('wf-1');
    expect(proof.stepHashes.length).toBe(2);
    // Should be sorted by stepId: step-1, then step-2 -> '0x1111', '0x2222'
    expect(proof.stepHashes[0]).toBe('0x1111');
    expect(proof.stepHashes[1]).toBe('0x2222');
    expect(proof.merkleRoot).toBeDefined();
  });

  it('rejects mismatching workflow IDs', () => {
    const results: WorkflowStepResult[] = [
      {
        workflowId: 'wf-2',
        stepId: 'step-1',
        nodeId: 'node-1',
        stepHash: '0x1111',
        localProof: {
          version: '1.0',
          workflowId: 'wf-2',
          stepHashes: ['0x1111'],
          timestamp: 1000,
          chainId: 1,
          blockTag: { finalized: true }
        }
      }
    ];

    expect(() => aggregator.aggregateProofs('wf-1', results, 1, { finalized: true })).toThrow(WnodeDeterminismError);
  });
});

File: archive/root_legacy/wnode-sdk-go/mesh/proof_test.go
package mesh

import (
	"testing"
	"github.com/wnodeltd/wnode/wnode-sdk-go"
)

func TestMeshProofAggregator_AggregatesDeterministically(t *testing.T) {
	aggregator := NewMeshProofAggregator()

	results := []WorkflowStepResult{
		{
			WorkflowID: "wf-1",
			StepID:     "step-2",
			NodeID:     "node-2",
			StepHash:   "0x2222",
			LocalProof: sdk.ProofOfCompute{
				Version:    "1.0",
				WorkflowID: "wf-1",
				StepHashes: []string{"0x2222"},
				Timestamp:  1000,
				ChainID:    1,
				BlockTag:   sdk.BlockTag{Finalized: true},
			},
		},
		{
			WorkflowID: "wf-1",
			StepID:     "step-1",
			NodeID:     "node-1",
			StepHash:   "0x1111",
			LocalProof: sdk.ProofOfCompute{
				Version:    "1.0",
				WorkflowID: "wf-1",
				StepHashes: []string{"0x1111"},
				Timestamp:  1000,
				ChainID:    1,
				BlockTag:   sdk.BlockTag{Finalized: true},
			},
		},
	}

	proof, err := aggregator.AggregateProofs("wf-1", results, 1, sdk.BlockTag{Finalized: true})
	if err != nil {
		t.Fatalf("Expected no error, got: %v", err)
	}

	if proof.Version != "1.0" {
		t.Fatalf("Expected version 1.0, got %s", proof.Version)
	}

	if len(proof.StepHashes) != 2 {
		t.Fatalf("Expected 2 step hashes, got %d", len(proof.StepHashes))
	}

	if proof.StepHashes[0] != "0x1111" || proof.StepHashes[1] != "0x2222" {
		t.Fatalf("Expected hashes to be sorted by stepId, got %v", proof.StepHashes)
	}
}

func TestMeshProofAggregator_RejectsMismatchingWorkflowID(t *testing.T) {
	aggregator := NewMeshProofAggregator()

	results := []WorkflowStepResult{
		{
			WorkflowID: "wf-2",
			StepID:     "step-1",
			NodeID:     "node-1",
			StepHash:   "0x1111",
			LocalProof: sdk.ProofOfCompute{
				Version:    "1.0",
				WorkflowID: "wf-2",
				StepHashes: []string{"0x1111"},
				Timestamp:  1000,
				ChainID:    1,
				BlockTag:   sdk.BlockTag{Finalized: true},
			},
		},
	}

	_, err := aggregator.AggregateProofs("wf-1", results, 1, sdk.BlockTag{Finalized: true})
	if err == nil {
		t.Fatal("Expected error for mismatching workflow ID, got nil")
	}
}
